### PR TITLE
DI Bridge: Changing to work in function of Dependencies instead of Shared logic

### PR DESCRIPTION
**Pull request by tdopires**
The way it's done today, **di-bridge** contains definition for the shared logic of feature-api modules.
This approach makes a little harder if you want to build feature demo apps that just compile one feature module, since you would have to build a whole new CoreComponent without the previous visibility of whats needed.

With this PR, I'm changing the CoreComponent to work in function of de dependencies of the feature modules, renaming it to AppComponentDependencies.
The inspiration for it came from [PR#28](https://github.com/alex-tavella/MooV/pull/28) whilst I was thinking on a way of doing that approach in a more safe way (without reified types on `dependencies()` extensions)

Finally, with this new approach, to build a feature demo app you would just build a new _YourFeature_ComponentDependencies interface with the set of feature modules Dependencies you need. 
i.e.:
```
interface HomeComponentDependencies : HomeDependencies

interface HomeComponentDependenciesProvider {
    fun dependencies(): HomeComponentDependencies
}
```
Then create an Application that implements that `HomeComponentDependenciesProvider`

After that, you would need to also change some extension function calls on your feature, creating more specific ways to get you dependencies like the ones from DependencyExtensions.kt (this last part still needs some improvements)

```
        ...
import br.com.moov.dibridge.HomeDependencies
import br.com.moov.dibridge.dependencies as diBridgeDependencies

class HomeActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        val moviesNavigator = dependencies().moviesNavigator()
        ...
    }

    private fun dependencies(): HomeDependencies {
        return if (application is HomeComponentDependenciesProvider) {
            (application as HomeComponentDependenciesProvider).dependencies()
        } else {
            diBridgeDependencies()
        }
    }
}
```

### CODE REVIEW

File: di-bridge/src/main/java/br/com/moov/dibridge/MovieDetailComponent.kt [hunk 1]
```diff
@@ -19,23 +19,12 @@ import br.com.moov.bookmark.movie.BookmarkMovieUseCase
 import br.com.moov.bookmark.movie.UnBookmarkMovieUseCase
 import br.com.moov.core.ImageUrlResolver
 import br.com.moov.moviedetails.navigation.MovieDetailsNavigator
-import br.com.moov.movies.navigation.MoviesNavigator
 import retrofit2.Retrofit
 
-interface CoreComponent : FeatureApisProvider, CoreApisProvider
-
-interface CoreApisProvider {
+interface MovieDetailsDependencies {
```
Comment (alex-tavella-ifood) [inline di-bridge/src/main/java/br/com/moov/dibridge/MovieDetailComponent.kt]: File should be `MovieDetailsDependencies`, right?

**Review by alex-tavella-ifood — COMMENTED** [review]
The main problem I see with this is having to add extra code on the feature implementation in order to "select" which :app variant I'm choosing from.

Another concern I see is having access to the dependencies of features other than mine.

We should discuss this subject a little further.

File: di-bridge/src/main/java/br/com/moov/dibridge/MovieDetailComponent.kt (hunk 1 above)
Comment (thiago-oliveira-ifood) [reply di-bridge/src/main/java/br/com/moov/dibridge/MovieDetailComponent.kt]: right, let me fix that

**Comment by tdopires** [comment]
> The main problem I see with this is having to add extra code on the feature implementation in order to "select" which :app variant I'm choosing from.
> 
Yes, that is really a concern... but I think it's a small cost that you would just do once on the feature module... 
It seems ok to me, but maybe I'm just too afraid of runtime Exceptions and crashes 😅 

> Another concern I see is having access to the dependencies of features other than mine.
> 
We could try to solve that by having feature-specific modules on the di-bridge layer...
For instance, `:movies-di-bridge` to have `MoviesDependencies`, and then we would have a `:app-di-bridge` (which today is `:di-bridge`) and also `:movies` to know it... but that could explode in terms of granularity.
I think having generic Dependencies would be nice (NetworkDependencies, etc) and could solve that problem very well.

**Pull request closed without merging**